### PR TITLE
add support for 'firstProperty' in schema

### PR DESCRIPTION
**Pull request by joscol**
firstProperty is an array of properties that can appear first for the object.  It affects both auto-complete and validation.

This pull request also suppresses validation warnings for deprecated objects/properties

### CODE REVIEW

**Comment by vtbassmatt** [comment]
Love it. I'm not qualified to do an actual code review :) but consider this my approval of the concept.

File: language-service/src/parser/jsonParser.ts [hunk 1]
```diff
@@ -818,6 +829,31 @@ export class ObjectASTNode extends ASTNode {
 			});
 		}
 	}
+
+	protected validateBestMatch(schema: JSONSchema, validationResult: ValidationResult): void {
+		if (schema.firstProperty && schema.firstProperty.length) {
```
Comment (stephenmichaelf) [inline language-service/src/parser/jsonParser.ts]: Combine both if statements?

File: language-service/src/parser/jsonParser.ts [hunk 2]
```diff
@@ -818,6 +829,31 @@ export class ObjectASTNode extends ASTNode {
...
+			if (this.properties && this.properties.length) {
+				let firstProperty = this.properties[0];
```
Comment (stephenmichaelf) [inline language-service/src/parser/jsonParser.ts]: Add type to all of these, use const where possible.

File: language-service/src/parser/jsonParser.ts [hunk 3]
```diff
@@ -818,6 +829,31 @@ export class ObjectASTNode extends ASTNode {
...
+				let firstPropKey: string = firstProperty.key.value;
+				let contained: boolean = false;
+				schema.firstProperty.forEach(listProperty => {
+					if (listProperty == firstPropKey) {
+						contained = true;
```
Comment (stephenmichaelf) [inline language-service/src/parser/jsonParser.ts]: break?

File: language-service/src/parser/jsonParser.ts [hunk 4]
```diff
@@ -818,6 +829,31 @@ export class ObjectASTNode extends ASTNode {
...
+					}
+				});
+
+				if (!contained) {
+					validationResult.problems.push({
+						location: { start: firstProperty.start, end: firstProperty.end },
+						severity: ProblemSeverity.Error,
+						//TODO update message to allow for multiple options in firstProperty
```
Comment (stephenmichaelf) [inline language-service/src/parser/jsonParser.ts]: Resolve TODO?

File: language-service/src/services/yamlCompletion.ts [hunk 5]
```diff
@@ -223,22 +223,40 @@ export class YAMLCompletion {
 		});
 	}
 
+	private checkPropertyArray(properties: string[], property: string): boolean {
```
Comment (stephenmichaelf) [inline language-service/src/services/yamlCompletion.ts]: Can we rename the function? Not sure what it means to check.

File: language-service/src/services/yamlCompletion.ts [hunk 6]
```diff
@@ -223,22 +223,40 @@ export class YAMLCompletion {
...
+		if (!properties || !properties.length) {
+			return true;
+		}
+
+		let contained: boolean = false;
+		properties.forEach(listProperty => {
+			if (listProperty == property) {
+				contained = true;
```
Comment (stephenmichaelf) [inline language-service/src/services/yamlCompletion.ts]: Nit: could just return true here and return false if we get past the foreach.

File: language-service/src/services/yamlCompletion.ts [hunk 7]
```diff
@@ -223,22 +223,40 @@ export class YAMLCompletion {
...
-							collector.add({
-								kind: CompletionItemKind.Property,
-								label: key,
-								insertText: this.getInsertTextForProperty(key, propertySchema, addValue, separatorAfter),
-								insertTextFormat: InsertTextFormat.Snippet,
-								documentation: propertySchema.description || ''
-							});
+						//check for more than one propery because the placeholder will always be in the list
+						if (s.node.properties.length > 1 || this.checkPropertyArray(s.schema.firstProperty, key)) {
+							let propertySchema = schemaProperties[key];
```
Comment (stephenmichaelf) [inline language-service/src/services/yamlCompletion.ts]: Add const and type.

File: language-service/src/services/yamlCompletion.ts [hunk 8]
```diff
@@ -223,22 +223,40 @@ export class YAMLCompletion {
...
-						let propertySchema = schemaProperties[key];
-						if (!propertySchema.deprecationMessage && !propertySchema["doNotSuggest"]) {
-							collector.add({
-								kind: CompletionItemKind.Property,
-								label: key,
-								insertText: this.getInsertTextForProperty(key, propertySchema, addValue, separatorAfter),
-								insertTextFormat: InsertTextFormat.Snippet,
-								documentation: propertySchema.description || ''
-							});
+						//check for more than one propery because the placeholder will always be in the list
```
Comment (stephenmichaelf) [inline language-service/src/services/yamlCompletion.ts]: Can we add tests for this?

File: language-service/src/services/yamlCompletion.ts [hunk 9]
```diff
@@ -223,22 +223,40 @@ export class YAMLCompletion {
...
+						if (s.node.properties.length > 1 || this.checkPropertyArray(s.schema.firstProperty, key)) {
```
Comment (stephenmichaelf) [inline language-service/src/services/yamlCompletion.ts]: Are we missing an or for the case when firstProperty isn't set?

File: language-service/src/parser/jsonParser.ts [hunk 10]
```diff
@@ -818,6 +829,31 @@ export class ObjectASTNode extends ASTNode {
...
+				let firstProperty = this.properties[0];
+				let firstPropKey: string = firstProperty.key.value;
+				let contained: boolean = false;
+				schema.firstProperty.forEach(listProperty => {
+					if (listProperty == firstPropKey) {
+						contained = true;
+					}
+				});
+
+				if (!contained) {
```
Comment (stephenmichaelf) [inline language-service/src/parser/jsonParser.ts]: Can we add acceptance tests for this?

File: language-service/package.json [hunk 11]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "azure-pipelines-language-service",
   "description": "Azure Pipelines language service",
-  "version": "0.2.3",
+  "version": "0.2.4",
```
[the remark this answers is not in the document]
Comment (joscol) [reply language-service/package.json]: I thought we had to use a new version number every time we pushed to NPM and we're going to want to push after this

File: language-service/src/parser/jsonParser.ts [hunk 12]
```diff
@@ -263,15 +266,23 @@ export class ASTNode {
 			}
 		}
 
+		//TODO find a way to soften the appearance of validation warnings
```
[the remark this answers is not in the document]
Comment (joscol) [reply language-service/src/parser/jsonParser.ts]: #137

File: language-service/src/parser/jsonParser.ts (hunk 3 above)
Comment (joscol) [reply language-service/src/parser/jsonParser.ts]: break doesn't work with `forEach`.  I'll change the loop structure to allow early termination

File: language-service/src/services/yamlCompletion.ts (hunk 6 above)
Comment (joscol) [reply language-service/src/services/yamlCompletion.ts]: early returns make code harder to read

File: language-service/src/services/yamlCompletion.ts (hunk 7 above)
Comment (joscol) [reply language-service/src/services/yamlCompletion.ts]: I will make it `const`.  `getMatchingSchemas` is an untyped mess that I will save for another day to fix.

File: language-service/src/services/yamlCompletion.ts (hunk 9 above)
Comment (joscol) [reply language-service/src/services/yamlCompletion.ts]: `checkPropertyArray` returns true for a null, undefined or empty list

File: language-service/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (stephenmichaelf) [reply language-service/package.json]: Yes, I just meant 0.3.0 :)

File: language-service/src/services/yamlCompletion.ts (hunk 8 above)
Comment (joscol) [reply language-service/src/services/yamlCompletion.ts]: Tests will be added later (see e-mail for details why)

File: language-service/package.json (hunk 11 above)
[the remark this answers is not in the document]
Comment (joscol) [reply language-service/package.json]: Oh, sorry.  I changed it to 0.3.0

**Pull request merged**